L112-node-server-interceptors.md

```diff
@@ -0,0 +1,226 @@
+
+```ts
+interface ServerListener {
+  /**
+   * An inbound interception method called when metadata is received. After this event goes through the interception stack, the regular method handler is called.
+   */
+  onReceiveMetadata?: (metadata: Metadata, next: (metadata: Metadata) => void): void;
+  /**
+   * An inbound interception method called when each message is received.
+   */
+  onReceiveMessage?: (message: any, next: (message: any) => void): void;
+  /**
+   * An inbound interception method called when the inbound message stream ends.
+   */
+  onReceiveHalfClose?: (next: () => void): void;
+  /**
+   * An inbound interception method called when the request ends for any reason, including when cancelled by the remote end.
+   */
+  onCancel?: (next: () => void): void;
+}
+```
+
+A `ServerListenerBuilder` will be provided to easily construct a listener:
+
+```javascript
+const listener = (new ServerListenerBuilder())
+      .withOnReceiveMetadata(function(metadata, next) {
+          logger.log(metadata);
+          next(metadata);
+      })
+      .withOnReceiveMessage(function(message, next) {
+          logger.log(message);
+          next(message);
+      })
+      .build();
+```
+
+#### Internal Errors and Exceptions
+
+Exceptions are not handled by the interceptor framework, we expect
+interceptor authors to either handle the exceptions their interceptors
+can throw or assume the Node process will crash on exceptions. This is in
+accordance with the [Joyent guidance](https://www.joyent.com/node-js/production/design/errors)
+to only catch exceptions which are known to be operational errors.
+
+In the case of an error during the execution of an interceptor,
+interceptor authors have the following options:
+1) Short-circuiting the call and returning a gRPC error to the client's consumer.
+2) Throwing an exception which the client's consumer will have to catch
+or crash on.
+3) Doing nothing and continuing with the call.
+
+### Use
+
+The `Server` constructor's `options` parameter will accept a new option `interceptors`, which is an array of interceptor functions.
+
+
+#### Interceptor nesting
+
+Interceptors nest in the order provided. Providing an interceptors array like `[interceptorA, interceptorB, interceptorC]` will produce this execution graph:
+
+```
+underlying ServerCall created ->
+  interceptorA call ->
+    interceptorB call ->
+      interceptorC call ->
+        start called ->
+      interceptorC start ->
+    interceptorB start ->
+  interceptorA start ->
+underlying ServerCall start ->
+  interceptorA onReceiveMetadata ->
+    interceptorB onReceiveMetadata ->
+      interceptorC onReceiveMetadata ->
+        method handler invocation ->
+      interceptorC outbound ->
+    interceptorB outbound ->
+  interceptorA outbound ->
+underlying ServerCall ->
+  interceptorA inbound ->
+    interceptorB inbound ->
+      interceptorC inbound ->
+        method handler
+```
+
+## Rationale
```

Comment:
Could you add a section on how closely (or not) the above APIs match the client-side interceptor design & apis?

Reply:
You can see the client interceptor APIs in https://github.com/grpc/proposal/blob/master/L5-node-client-interceptors.md. They are as close as possible, but they're not identical because there are inherent differences in client and server functionality. I said that the design "mirrors" the client interceptor design in the abstract, in the sense that client send operations generally correspond to server receive operations, and vice versa.

Reply:
I added a section to show the correspondence.